docs/source/intro.rst

```diff
@@ -805,12 +819,13 @@ increasing chance of halting, but that may run forever!
 
 .. code:: python
 
-    inside_loop = Program(H(0)).measure(0, 1)
-
-    p = Program().inst(X(0)).while_do(1, inside_loop)
+    p = Program()
+    ro = p.declare('ro', 'BIT', 1)
+    inside_loop = Program(H(0)).measure(0, ro[0])
+    p.inst(X(0)).while_do(0, inside_loop)
 
-    # Run and check register [1]
-    print(quantum_simulator.run(p, [1]))
+    qc = get_qc('9q-square-qvm')
+    print (qc.run(qc.compile(p)))
```

Comment:
This and the `if_else` example above gave me a QUILC compilation error. Running the `while_do` probabilistic halting problem example gave me the following traceback:
```
>>> print (qc.run(qc.compile(p)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/amy/Repositories/pyquil/pyquil/api/_error_reporting.py", line 234, in wrapper
    val = func(*args, **kwargs)
  File "/Users/amy/Repositories/pyquil/pyquil/api/_quantum_computer.py", line 224, in compile
    nq_program = self.compiler.quil_to_native_quil(program)
  File "/Users/amy/Repositories/pyquil/pyquil/api/_compiler.py", line 223, in quil_to_native_quil
    response = self._connection._quilc_compile(program, self.isa, self.specs)
  File "/Users/amy/Repositories/pyquil/pyquil/api/_base_connection.py", line 395, in _quilc_compile
    response = post_json(self.session, self.sync_endpoint + "/quilc", payload)
  File "/Users/amy/Repositories/pyquil/pyquil/api/_base_connection.py", line 58, in post_json
    raise parse_error(res)
pyquil.api._errors.QUILCError: Misplaced or illegal instruction in ProtoQuil program:
   X 0
>>>LABEL @START1
   JUMP-UNLESS @END2 ro[0]
```

The Quil instructions this code gave me were
```
>>> print(p.out())
DECLARE ro BIT[1]
X 0
LABEL @START1
JUMP-UNLESS @END2 ro[0]
H 0
MEASURE 0 ro[0]
JUMP @START1
LABEL @END2
```

Reply:
That's weird. I get the same Quil output, but it runs fine for me. What version of pyquil are you using? I have '2.0.0'.

Reply:
@msohaibalam, hmm I can't seem to get it to work with 2.0.0. 
```(new-pyquil) amy@amy-macbook:~/Repositories/amy/amy (grove)$ python pyquil-snippet.py
Running on pyQuil version 2.0.0

Running while-do example

/Users/amy/miniconda3/envs/new-pyquil/lib/python3.6/site-packages/pyquil/quilatom.py:621: UserWarning: Addr objects have been deprecated. Defaulting to memory region "ro". Use MemoryReference instead.
  warn("Addr objects have been deprecated. Defaulting to memory region \"ro\". Use MemoryReference instead.")
Traceback (most recent call last):
  File "pyquil-snippet.py", line 33, in <module>
    run_while_do_example()
  File "pyquil-snippet.py", line 26, in run_while_do_example
    print (qc.run(qc.compile(p)))
  File "/Users/amy/miniconda3/envs/new-pyquil/lib/python3.6/site-packages/pyquil/api/_error_reporting.py", line 234, in wrapper
    val = func(*args, **kwargs)
  File "/Users/amy/miniconda3/envs/new-pyquil/lib/python3.6/site-packages/pyquil/api/_quantum_computer.py", line 224, in compile
    nq_program = self.compiler.quil_to_native_quil(program)
  File "/Users/amy/miniconda3/envs/new-pyquil/lib/python3.6/site-packages/pyquil/api/_compiler.py", line 223, in quil_to_native_quil
    response = self._connection._quilc_compile(program, self.isa, self.specs)
  File "/Users/amy/miniconda3/envs/new-pyquil/lib/python3.6/site-packages/pyquil/api/_base_connection.py", line 395, in _quilc_compile
    response = post_json(self.session, self.sync_endpoint + "/quilc", payload)
  File "/Users/amy/miniconda3/envs/new-pyquil/lib/python3.6/site-packages/pyquil/api/_base_connection.py", line 58, in post_json
    raise parse_error(res)
pyquil.api._errors.QUILCError: Misplaced or illegal instruction in ProtoQuil program:
   X 0
>>>LABEL @START1
   JUMP-UNLESS @END2 ro[0]

QUILC returned the above error. This could be due to a bug in the server or a
bug in your code. If you suspect this to be a bug in pyQuil or Rigetti Forest,
then please describe the problem in a GitHub issue at:
    https://github.com/rigetticomputing/pyquil/issues
```
I suspect my outdated quilc and qvm binaries are to blame. I'll update them and report back.

Reply:
@amyfbrown I updated pyquil to version `'2.0.0.dev2'` (using `pip install --upgrade pyquil`), deleted/reinstalled qvm and quilc, with versions `1.0.1` and `1.0.2` respectively. I am still unable to recover the error msg you are seeing, and the code seems to run without complaint. Note that with the changes from #674, one would need to change the line
```p.inst(X(0)).while_do(0, inside_loop)```
to
```p.inst(X(0)).while_do(ro[0], inside_loop)```
and also avoid the Warning Message `UserWarning: Addr objects have been deprecated. Defaulting to memory region "ro". Use MemoryReference instead.` in the process.

This also begs the question @mpharrigan, should the code samples in this docs page be consistent with the version of pyquil you would get from running `pip install pyquil` (or `pip install --upgrade pyquil`), currently version `'2.0.0.dev2'`, OR the version of pyquil you would get from running `pip install -e .` using the latest code on the github master branch, currently `'2.1.0.dev0'`?

I believe this is the last remaining issue for this PR, so it'd be great if we could resolve this soon.

Reply:
docs as merged to `master` should match code on `master` :)

Reply:
Ah whoops, in that case I need to make a small update.

Reply:
@mpharrigan Thanks for merging this PR. Created a fresh PR #695 that makes the `while_do(..)` statement compatible with the current code on master.

Reply:
great. That actually works out quite nicely. I backported these docs changes to the `release-2.0` branch so we can update the `/stable/` docs for pyQuil 2.0 (if we want to) before the release of 2.1. I *won't* backport #695 since it *only* applies to v2.1

In the future this won't be as confusing because we'll keep the docs up-to-date as we implement features instead of releasing the software and *then* writing all the docs :)